vm/devices/hyperv_ic/src/shutdown.rs

```diff
@@ -39,10 +45,79 @@ use zerocopy_helpers::FromBytesExt;
 /// A shutdown IC device.
 #[derive(InspectMut)]
 pub struct ShutdownIc {
+    #[inspect(skip)]
+    disconnected_state_runner: mesh::Sender<ShutdownIcStateTransition>,
+    state: Option<ShutdownIcState>,
+}
+
+/// An RPC request to get/set shutdown state.
+#[derive(MeshPayload)]
+enum ShutdownIcStateTransition {
+    /// Process state for a disconnected vmbus channel.
+    Set(ShutdownIcState),
+    /// Retrieve state for a connected vmbus channel.
+    Get(Rpc<(), ShutdownIcState>),
+}
+
+// Current state of the shutdown IC device.
+#[derive(Inspect, MeshPayload)]
+struct ShutdownIcState {
     #[inspect(skip)]
     recv: mesh::Receiver<ShutdownRpc>,
     #[inspect(skip)]
     wait_ready: Vec<Rpc<(), ()>>,
+    #[inspect(skip)]
+    wait_not_ready: Vec<Rpc<(), ()>>,
+}
+
+async fn disconnected_task(
```

Comment:
This adds some annoying complexity to the device. Let's use a little more mesh cleverness to avoid needing this.

Here's what I'm imagining. Today, we have a model where the shutdown device has a mesh receiver for RPCs (WaitReady, WaitNotReady, Shutdown). The "client" side uses the sending side of this channel to send shutdown requests whenever it wants.

Instead, make WaitReady return a new sender to use just for shutdown requests. The device will hang on to the receiver side as long as the vmbus channel is open. But once the vmbus channel is closed, it will drop the receiver side, and so subsequent RPCs will automatically fail with channel closed errors. That way, you don't have to separately handle these RPCs via this task.

Reply:
I'd also suggest that WaitReady should return a OneshotReceiver that gets closed as well, rather than add the new WaitNotReady message. The caller can keep this around and wait on it.

(Ideally we'd fix mesh so that you can just wait on a sender, to be notified when the receiving end closes. I'll add that to the mesh backlog.)